code/game/mecha/combat/durand.dm

```diff
@@ -20,3 +34,178 @@
+	else
+		. = ..()
+
+////////////////////////////
+///// Shield processing ////
+////////////////////////////
+
+/**An object to take the hit for us when using the Durand's defense mode.
+It is spawned in during the durand's initilization, and always stays on the same tile.
+Normally invisible, until defense mode is actvated. When the durand detects an attack that should be blocked, the
+attack is passed to the shield. The shield takes the damage, uses it to calculate charge cost, and then sets its
+own integrity back to max. Shield is automatically dropped if we run out of power or the user gets out.*/
+
+/obj/durand_shield //projectiles get passed to this when defense mode is enabled
+	name = "defense grid"
+	icon = 'icons/mecha/durand_shield.dmi'
+	icon_state = "shield_null"
+	invisibility = INVISIBILITY_MAXIMUM //no showing on right-click
+	pixel_y = 4
+	max_integrity = 10000
+	obj_integrity = 10000
+	var/obj/mecha/combat/durand/chassis ///Our link back to the durand
+	var/switching = FALSE ///To keep track of things during the animation
+
+/obj/durand_shield/Initialize()
+	. = ..()
+	RegisterSignal(src, COMSIG_MECHA_ACTION_ACTIVATE, .proc/activate)
+
+/obj/durand_shield/Destroy()
+	if(chassis)
+		chassis.shield = null
+	. = ..()
+
+/**Handles activating and deactivating the shield. This proc is called by a signal sent from the mech's action button
+and relayed by the mech itself. The "forced" variabe, signal_args[1], will skip the to-pilot text and is meant for when
+the shield is disabled by means other than the action button (like running out of power)*/
+
+/obj/durand_shield/proc/activate(datum/source, var/datum/action/innate/mecha/mech_defense_mode/button, list/signal_args)
+	if(!chassis || !chassis.occupant)
+		return
+	if(switching && !signal_args[1])
+		return
+	if(!chassis.defense_mode && (!chassis.cell || chassis.cell.charge < 100)) //If it's off, and we have less than 100 units of power
+		chassis.occupant_message("<span class='warn'>Insufficient power; cannot activate defense mode.</span>")
+		return
+	switching = TRUE
+	chassis.defense_mode = !chassis.defense_mode
+	chassis.defense_action.button_icon_state = "mech_defense_mode_[chassis.defense_mode ? "on" : "off"]" //This is backwards because we haven't changed the var yet
+	if(!signal_args[1])
+		chassis.occupant_message("<span class='notice'>Defense mode [chassis.defense_mode?"enabled":"disabled"].</span>")
+		chassis.log_message("User has toggled defense mode -- now [chassis.defense_mode?"enabled":"disabled"].", LOG_MECHA)
+	else
+		chassis.log_message("defense mode state changed -- now [chassis.defense_mode?"enabled":"disabled"].", LOG_MECHA)
+	chassis.defense_action.UpdateButtonIcon()
+
+	if(chassis.defense_mode)
+		invisibility = 0
+		flick("shield_raise", src)
+		playsound(src, 'sound/mecha/mech_shield_raise.ogg', 50, FALSE)
+		set_light(l_range = MINIMUM_USEFUL_LIGHT_RANGE	, l_power = 5, l_color = "#00FFFF")
+		sleep(3)

```

Comment:
avoid using sleep at all.
try setting icon state before the flick

Reply:
The sleep command allows for a few fluff effects, like the shield not blocking damage until it's fully up and emitting light until the very last frame of the animation. It also handles setting the shield's invisibility when being dropped, so that disabled shields don't show up on the right-click menu.

I could probably replace it with timers and extra procs, but are sleeps really that bad of an issue?

Reply:
I looked more into it, and this is fine.
I was under the impression that sleeps are always bad news, but they can be safe in some cases.